store/localcachelayer/user_layer.go

```diff
@@ -158,5 +156,5 @@ func (s LocalCacheUserStore) Get(id string) (*model.User, *model.AppError) {
 		return nil, model.NewAppError("SqlUserStore.Get", "store.sql_user.get.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 	s.rootStore.doStandardAddToCache(s.rootStore.userProfileByIdsCache, id, user)
-	return user, nil
+	return user.DeepCopy(), nil
```

Comment:
Same here?

Reply:
Same thing here. We are storing it in the cache so we should make a copy. Of course we could cache the copy and return the data fetched directly from the store but that's the same thing in the end. I tried to keep consistent in both functions and always return a copy while caching the store value.

Reply:
For now we have an 'in memory' cache, but probably soon we will have a redis based cache, which doesn't need this DeepCopy at all, is there any way to move it to the specific cache implementation, in this case to the lru?

Makes sense to do it only for one specific case? Should we always return a shallow copy, at least, from the lru perspective?

Reply:
I'm now thinking about it, and at least we should be shallow coping on insert, and shallow coping on return, because if not, you can add something to the cache, and modify it later or you can get something from the cache and modify it, in both cases you are modifying the cached version.

Reply:
> For now we have an 'in memory' cache, but probably soon we will have a redis based cache, which doesn't need this DeepCopy at all, is there any way to move it to the specific cache implementation, in this case to the lru?
> 
> Makes sense to do it only for one specific case? Should we always return a shallow copy, at least, from the lru perspective?

Absolutely, we should probably return *at least* a shallow copy from any cache related function. This PR was more about giving a quick fix to the reported issues but there's definitely a need to think about a more long term solution and doing it inside the LRU implementation sounds like the better idea.

Reply:
> I'm now thinking about it, and at least we should be shallow coping on insert, and shallow coping on return, because if not, you can add something to the cache, and modify it later or you can get something from the cache and modify it, in both cases you are modifying the cached version.

Yes, that's the problem we are facing that is causing all these weird bugs.

Reply:
FWIW I've got this down as something we need to fix properly next time we have an opportunity to break binary compatibility. Epic here: https://mattermost.atlassian.net/browse/MM-21733